plugins/destination/firehose/client/write.go

```diff
@@ -0,0 +1,133 @@
+	if err != nil {
+		c.logger.Error().Err(err).Msg("invalid firehose stream ARN")
+		return err
+	}
+	resource := strings.Split(parsedARN.Resource, "/")
+	if len(resource) != 2 {
+		c.logger.Error().Err(err).Msg("invalid firehose stream ARN")
+		return fmt.Errorf("invalid firehose stream ARN")
+	}
+	recordsBatchInput := &firehose.PutRecordBatchInput{
+		DeliveryStreamName: aws.String(resource[1]),
+	}
+	batchSize := 0
+
+	for resource := range res {
+		table := tables.Get(resource.TableName)
+		if table == nil {
+			panic(fmt.Errorf("table %s not found", resource.TableName))
+		}
+
+		jsonObj := make(map[string]any, len(table.Columns)+1)
+		for i := range resource.Data {
+			jsonObj[table.Columns[i].Name] = resource.Data[i]
+		}
+		// Add table name to the json object
+		// TODO: This should be added to the SDK so that it can be used for other plugins as well
+		jsonObj["_cq_table_name"] = table.Name
+		b, err := json.Marshal(jsonObj)
+		if err != nil {
+			return err
+		}
+		dst := &bytes.Buffer{}
+		err = json.Compact(dst, b)
+		if err != nil {
+			return err
+		}
+		if len(dst.Bytes()) > MAX_RECORD_SIZE_BYTES {
+			c.logger.Warn().Msgf("skipping record because it is too large: %s", string(b))
+			continue
+		}
+
+		// If adding this record would exceed the batch size, send the batch
+		if len(dst.Bytes())+batchSize > min(c.spec.BatchSizeBytes, MAX_BATCH_SIZE_BYTES) {
+			err := c.sendBatch(ctx, recordsBatchInput, 0)
+			if err != nil {
+				return err
+			}
+			recordsBatchInput.Records = nil
+			batchSize = 0
+		}
+
+		recordsBatchInput.Records = append(recordsBatchInput.Records, types.Record{
+			Data: dst.Bytes(),
+		})
+		// Store a running total of the batch size
+		batchSize += len(dst.Bytes())
+
+		// Send the batch if it is full
+		if len(recordsBatchInput.Records) == min(c.spec.BatchSize, MAX_BATCH_RECORDS) {
+			err := c.sendBatch(ctx, recordsBatchInput, 0)
+			if err != nil {
+				return err
+			}
+			// Reset the batch
+			recordsBatchInput.Records = nil
+			batchSize = 0
+		}
+	}
+	// Send the last batch
+	return c.sendBatch(ctx, recordsBatchInput, 0)
+}
+
+func (c *Client) sendBatch(ctx context.Context, recordsBatchInput *firehose.PutRecordBatchInput, count int) error {
+	if count == *c.pluginSpec.MaxRetries {
+		return fmt.Errorf("max retries reached")
+	}
+	if recordsBatchInput == nil || len(recordsBatchInput.Records) == 0 {
+		return nil
+	}
+	time.Sleep(time.Duration(count) * time.Second)
+	resp, err := c.firehoseClient.PutRecordBatch(ctx, recordsBatchInput)
+	if err != nil {
+		c.logger.Error().Err(err).Msg("failed to write to firehose")
+		return err
+	}
+	retryRecords := getFailedRecords(recordsBatchInput, resp)
+	atomic.AddUint64(&c.metrics.Writes, uint64(len(recordsBatchInput.Records)-len(retryRecords.Records)))
+	return c.sendBatch(ctx, retryRecords, count+1)
+}
+
+func getFailedRecords(recordsBatchInput *firehose.PutRecordBatchInput, resp *firehose.PutRecordBatchOutput) *firehose.PutRecordBatchInput {
```

Comment:
Let's not do that. either a batch failed or succeeded and we log it. this will introduce a behaviour we had in v0 that we have no idea why something taking too long is it because we used `copyfrom`, `insertbatch` or `insert one by one` because the batch failed

Reply:
In general a failed batch should rarely happen so we shouldn't have a code handling that as most of the validation is hapening on the SDK side

Reply:
For Batch requests the SDK doesn't handle retries for individual records that are rate limited... This is the case for Kinesis Firehose, Kinesis Data streams, SQS, and DynamoDB... So it is actually quite easy to hit the default rate limit...

Reply:
The SDK will only retry if the request returns an retryable error... In the cases of Batch endpoints the request can return a 200 status code, but every single record could have been rate limited and therefore not inserted...